migrations/Version20190109000000.php

```diff
@@ -30,33 +38,44 @@ public function postUp( Schema $schema ) {
 		$pageCount = (int)ceil( $entityCount / self::DB_QUERY_PAGE_SIZE );
 		echo $entityCount . ' entries to be updated.' . PHP_EOL;
 		for ( $page = 0; $page < $pageCount; $page++ ) {
-			echo 'Page ' . ( $pageCount + 1 ) . ' / ' . $pageCount . PHP_EOL;
+			echo 'Page ' . ( $page + 1 ) . ' / ' . $pageCount . PHP_EOL;
 			$query = '';
 			foreach ( $this->getCurrentPage() as $addressEntry ) {
 				$data = $addressEntry['donation_data'] ?? $addressEntry['membership_data'];
+				if ( $data === null ) {
+					// Entries with no data are deleted from the AddressChange table and the foreign key reference is removed
+					$query .= $this->getAddressChangeDeleteQueries( (int) $addressEntry['id'] );
+					continue;
+				}
 				$type = $this->getAddressType( $data );
-				$query .= 'UPDATE address_change ( address_type ) VALUES ( "' . $type . '" );';
+				$query .= 'UPDATE address_change SET address_type = "' . $type . '" WHERE id = "' . $addressEntry['id'] . '";';
 			}
 			if ( $query !== '' ) {
 				$stmt = $this->connection->prepare( $query );
 				$stmt->execute();
 				$stmt->closeCursor();
 				$this->connection->commit();
+				$this->connection->beginTransaction();
```

Comment:
I thought this class is auto generated code. Is it not?

Reply:
It is, but the `beginTransaction()` call is added by us. I just looked at Doctrine's docs.
https://www.doctrine-project.org/projects/doctrine1/en/latest/manual/transactions.html#introduction
They describe the order of execution as it is in MySQL.

Reply:
Yeah, the thing was, if I started with beginTransaction, the performance was horrible and it would have taken hours to finish the migration script. By placing the beginTransaction at the end, all queries were performed correctly and the complete migration was done in about 10 minutes on my local machine. I know it does not make sense and I tried to debug it, but all "proper" attempts ended up being super slow. I assume that it has something to do with how Doctrine does transaction nesting as described here:

https://www.doctrine-project.org/projects/doctrine-dbal/en/2.9/reference/transactions.html#transaction-nesting

I tried using two "beginTransactions" but that also did not improve the performance. I forgot to leave a comment in the code to describe this, which is why this is indeed confusing without this background knowledge, though I had talked about this with Gabriel.

I now saw that @gbirke has changed this, I think I attempted his solution as well and it was super slow for me, but maybe I did something differently, I have not tried his implementation yet.